packages/backend-storage/src/factory.ts

```diff
@@ -46,7 +55,143 @@ class AmplifyStorageFactory
+ */
+export class StorageOutputsAspect implements IAspect {
+  isStorageProcessed = false;
+  node: IConstruct;
+  outputStorageStrategy;
+  /**
+   * Constructs a new instance of the StorageValidator class.
+   */
+  constructor(
+    outputStorageStrategy: BackendOutputStorageStrategy<StorageOutput>
+  ) {
+    this.outputStorageStrategy = outputStorageStrategy;
+  }
+
+  /**
+   * Visit the given node.
+   * If the node is an AmplifyStorage construct, we will traverse its siblings in the same stack
+   * @param node The node to visit.
+   */
+  public visit(node: IConstruct): void {
+    if (!(node instanceof AmplifyStorage)) {
+      return;
+    }
+    const storageInstances = Stack.of(node).node.children.filter(
+      (el) => el instanceof AmplifyStorage
+    );
+    const storageCount = storageInstances.length;
+
+    /**
+     * only store traverse the siblings once to store the outputs,
+     * otherwise there will be the an error -
+     * "there is already a construct with name 'storageRegion'"
+     */
+    if (this.isStorageProcessed) {
+      return;
+    }
+    this.isStorageProcessed = true;
+    this.node = node;
+    let defaultStorageFound = false;
+
+    Stack.of(node).node.children.forEach((child) => {
+      if (!(child instanceof AmplifyStorage)) {
+        return;
+      }
+      if (child.isDefault && !defaultStorageFound) {
+        defaultStorageFound = true;
+      } else if (child.isDefault && defaultStorageFound) {
+        throw new AmplifyUserError('MultipleDefaultStorageError', {
+          message: `More than one default storage set in the Amplify project.`,
+          resolution:
+            'Remove `isDefault: true` from all `defineStorage` calls except for one in your Amplify project.',
+        });
+      }
+    });
+    /*
+     * If there is no default bucket set and there is only one bucket,
+     * we need to set the bucket as default.
+     */
+    if (!defaultStorageFound && storageCount === 1) {
+      this.storeOutput(
+        this.outputStorageStrategy,
+        true,
+        node.name,
+        node.resources.bucket.bucketName
+      );
+    } else if (!defaultStorageFound && storageCount > 1) {
+      throw new AmplifyUserError('NoDefaultStorageError', {
+        message: 'No default storage set in the Amplify project.',
+        resolution:
+          'Add `isDefault: true` to one of the `defineStorage` calls in your Amplify project.',
+      });
+    } else {
+      Stack.of(node).node.children.forEach((child) => {
+        if (!(child instanceof AmplifyStorage)) {
+          return;
+        }
+        this.storeOutput(
+          this.outputStorageStrategy,
+          child.isDefault,
+          child.name,
+          child.resources.bucket.bucketName
+        );
+      });
+    }
+  }
+
+  private storeOutput = (
+    outputStorageStrategy: BackendOutputStorageStrategy<StorageOutput> = new StackMetadataBackendOutputStorageStrategy(
```

Comment:
Do we need this default initialization?

Reply:
If it needed the default initialization when it was in the construct.ts, why don't it need it when it's moved here?

Reply:
because now it's private and all the call sites are passing a fully formed object.

Reply:
Just checked. The only place we call it is here https://github.com/aws-amplify/amplify-backend/blob/9d28f1ac43e2bbbcbc9e0d2e2912e2094e9087c9/packages/backend-storage/src/storage_container_entry_generator.ts#L37, and it's not optional, so yea, we don't need the default value.

Reply:
Done 1a7f0950eb